src/monattack.cpp

```diff
@@ -2255,6 +2255,172 @@ void mattack::tentacle(monster *z, int index)
     g->u.check_dead_state();
 }
 
+
+void mattack::pull_enemy(monster *z, int index)
+{
+    int t;
+    int t2 = 0;
+    Creature *target = z->attack_target();
+    monster *zz = dynamic_cast<monster*>(target);
+    if( target == nullptr || rl_dist( z->pos(), target->pos() ) > 3
+            || rl_dist( z->pos(), target->pos() ) <= 1 || !z->sees(*target, t)) {
+        return;
+    }
+
+    player *foe = dynamic_cast< player* >( target );
+    std::vector<tripoint> line = line_to( z->pos(), target->pos(), t, t2 );
+    bool seen = g->u.sees( *z );
+
+    z->reset_special(index); // Reset timer
+    z->moves -= 150;
+
+    for (auto &i : line){
+        const ter_t &terrain = g->m.ter_at(i);
+        //Player can't be pulled though bars
+        if (terrain.movecost == 0 ){
+            add_msg( _("The %s flings its arms at you, but they bounce off the %s"), z->name().c_str(), terrain.name.c_str() );
+            return;
+        }
+    }
+    bool uncanny = target->uncanny_dodge();
+
+    if( uncanny || dodge_check(z, target) ) {
+        z->moves -=200;
+        if( foe != nullptr ) {
+            if( seen ) {
+                auto msg_type = foe == &g->u ? m_warning : m_info;
+                foe->add_msg_player_or_npc( msg_type, _("The %s's arms fly out at you, but you dodge!"),
+                                                      _("The %s's arms fly out at <npcname>, but they dodge!"),
+                                            z->name().c_str() );
+            }
+            if( !uncanny ) {
+                target->on_dodge( z, z->type->melee_skill * 2 );
```

Comment:
`on_dodge` should be invoked regardless of creature type (player/monster/npc).
You could just remove the entire if and just have the `add_msg_player_or_npc` cover all cases. It uses the second message if invoked on a monster and replaces `<npcname>` with monster's name. You don't need to add check for `seen`, because `add_msg_player_or_npc` does it automatically.